### PR TITLE
When encoding using an existing tcf string, added changes to not igno…

### PR DESCRIPTION
**Pull request by srinivas81**
…re the publisher restrictions.

### CODE REVIEW

File: iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java [hunk 1]
```diff
@@ -407,6 +409,13 @@ public Builder(TCString tcString) {
                 purposeOneTreatment = tcString.getPurposeOneTreatment();
                 publisherCC = tcString.getPublisherCC();
                 vendorLegitimateInterest = BitSetIntIterable.newBuilder(tcString.getVendorLegitimateInterest());
+                for (PublisherRestriction pr : tcString.getPublisherRestrictions()) {
+                    PublisherRestrictionEntry pre = PublisherRestrictionEntry.newBuilder()
```
[the remark this answers is not in the document]
Comment (srinivas81) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java:413]: Yes absolutely, I didn't know the history of these classes, so didn't want to merge it. I will add a new commit soon. Thanks.

[the remark this answers is not in the document]
Comment (srinivas81) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java:413]: @laktech Do you like to keep the newBuilder static class provided in PublisherRestrictionEntry (in iabtcf-encoder package) and merge it or would you want to keep it simple and just rely on the constructor of PublisherRestriction class (in iabtcf-decoder package) for making new objects.

[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java:413]: great question... the builder pattern may be helpful in the encoder as it supports things like incrementally adding vendor ids to the publisher restriction. we want to publisher restriction to be immutable.

It ultimately comes down to, do we use the builder and if we do it's probably worthwhile to port it. I suspect we do. what do you think? probably doesn't hurt to port the builder pattern to the PublisherRestriction object in iabtcf-decoder. it's not much code.

[the remark this answers is not in the document]
Comment (srinivas81) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java]: @laktech Do you want to check the second commit if that seems reasonable? Thanks.

[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java]: hmm... one thing I didn't realize by this is that we've now introduced a dependency on the decoder for encoding purposes.

maybe we should leave them separate?

[the remark this answers is not in the document]
Comment (srinivas81) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java]: There is already a dependency on the decoder for encoding purposes.. I think its to use the IntIterable and few other classes.

[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/main/java/com/iabtcf/encoder/TCStringEncoder.java]: You're right, we're good then.

File: iabtcf-decoder/src/main/java/com/iabtcf/v2/PublisherRestriction.java [hunk 2]
```diff
@@ -32,11 +33,15 @@
     private final RestrictionType restrictionType;
     private final IntIterable vendorIds;
 
-    public PublisherRestriction(
+    private PublisherRestriction(
             int purposeId, RestrictionType restrictionType, IntIterable vendorIds) {
         Objects.requireNonNull(vendorIds);
         Objects.requireNonNull(restrictionType);
 
+        if (purposeId <= 0) {
```
Comment (laktech) [inline iabtcf-decoder/src/main/java/com/iabtcf/v2/PublisherRestriction.java:41]: can you reference the spec for this? also, can you move this check down to the builder?

Comment (srinivas81) [reply iabtcf-decoder/src/main/java/com/iabtcf/v2/PublisherRestriction.java]: I am not able to find any spec for this check. The spec just defines purposes. I just copied this check from the existing PublisherRestrictionEntry class

File: iabtcf-encoder/src/test/java/com/iabtcf/encoder/TCStringV2EncoderTest.java [hunk 3]
```diff
@@ -425,4 +425,13 @@ public void testToTCString() {
...
         assertNotEquals(tcStr, b.toTCString());
     }
+
+    @Test
+    public void testEncodingUsingTcfStringWithPublisherRestrictions() {
+        String tcStr =
+            "CO9ZBNoO9ZBNoAOACAENAnCsAP_AAH_AACiQFvNR6RBUJWjjMCAZg7BAAQQLIBADAAwAhAQIAAAAQIAEoBgGKEEQnBUgAAAACAAgAAYAAAAkFCAAAQAAAAAAgQIMQECICwBAAJAAgEEQQwAQAARMAgFSEQAAREwAEC0CARAAqAJAhCAAEgRAgAIAEAAgAgAAIQAgAAAAAAAACUAQBFCAAAAgAAAAAAAAQAAEAAAAAAAAAAAAAAAAAAIAEAAAABoAAAAAAQCAAAQAIAAACAAAIAAAAAAIAYaAQACoAJkAVQBWAFIANIAcQCbAE7CIBAAKgAmQBVAFYAUgArgBxAJsATsKgDAAqACYAI4AUgA4gE2A";
+        TCString tcString = TCString.decode(tcStr);
+        TCStringEncoder.Builder tcStringEncoder = TCStringEncoder.newBuilder(tcString);
+        assertEquals(tcStr, tcStringEncoder.encode());
```
[the remark this answers is not in the document]
Comment (srinivas81) [reply iabtcf-encoder/src/test/java/com/iabtcf/encoder/TCStringV2EncoderTest.java]: Done, please review

File: iabtcf-decoder/src/main/java/com/iabtcf/v2/PublisherRestriction.java (hunk 2 above)
Comment (srinivas81) [reply iabtcf-decoder/src/main/java/com/iabtcf/v2/PublisherRestriction.java]: Moved it to the build method.

File: iabtcf-encoder/src/test/java/com/iabtcf/encoder/TCStringV2EncoderTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (laktech) [reply iabtcf-encoder/src/test/java/com/iabtcf/encoder/TCStringV2EncoderTest.java]: sorry i wasn't clear here. It's difficult to determine what this test is doing with that TCString. The ask here is to make this test explicit and clear.

Can you use the encoder to create a TCString object, call it tcStringUnderTest, that has publisher restrictions and then invoke `TCStringEncoder.newBuilder(tcStringUnderTest.toTCString())`? And then confirm that the publisher restrictions propagated properly?

[the remark this answers is not in the document]
Comment (srinivas81) [reply iabtcf-encoder/src/test/java/com/iabtcf/encoder/TCStringV2EncoderTest.java]: Yeah, I think I understood you, so I added a new test case doing what you described. There seems to be two ways to encode a string, either by passing an existing String or by passing a builder, so this specific one I decided to start with the string.  If you think we don't need both the test cases, then I can delete this one.

**Review by laktech — APPROVED** [review]
looks good thanks!

**Pull request merged**